examples/generate.sh

```diff
@@ -0,0 +1,106 @@
+
+set -o errexit
+set -o nounset
+
+# Directories.
+SOURCE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+OUTPUT_DIR=${OUTPUT_DIR:-${SOURCE_DIR}/_out}
+ENVSUBST=${ENVSUBST:-envsubst}
+CLUSTERAWSADM=${CLUSTERAWSADM:-clusterawsadm}
+
+# Cluster name.
+export CLUSTER_NAME="${CLUSTER_NAME:-test1}"
+
+# Machine settings.
+export CONTROL_PLANE_MACHINE_TYPE="${CONTROL_PLANE_MACHINE_TYPE:-t2.medium}"
+export NODE_MACHINE_TYPE="${CONTROL_PLANE_MACHINE_TYPE:-t2.medium}"
+export SSH_KEY_NAME="${SSH_KEY_NAME:-default}"
+
+# Outputs.
+COMPONENTS_CLUSTER_API_GENERATED_FILE=${SOURCE_DIR}/provider-components/provider-components-cluster-api.yaml
+COMPONENTS_KUBEADM_GENERATED_FILE=${SOURCE_DIR}/provider-components/provider-components-kubeadm.yaml
+COMPONENTS_AWS_GENERATED_FILE=${SOURCE_DIR}/provider-components/provider-components-aws.yaml
+
+PROVIDER_COMPONENTS_GENERATED_FILE=${OUTPUT_DIR}/provider-components.yaml
+CLUSTER_GENERATED_FILE=${OUTPUT_DIR}/cluster.yaml
+CONTROLPLANE_GENERATED_FILE=${OUTPUT_DIR}/controlplane.yaml
+
+# Overwrite flag.
+OVERWRITE=0
+
+SCRIPT=$(basename "$0")
+while test $# -gt 0; do
+        case "$1" in
+          -h|--help)
+            echo "$SCRIPT - generates input yaml files for Cluster API on aws"
+            echo " "
+            echo "$SCRIPT [options]"
+            echo " "
+            echo "options:"
+            echo "-h, --help                show brief help"
+            echo "-f, --force-overwrite     if file to be generated already exists, force script to overwrite it"
+            exit 0
+            ;;
+          -f)
+            OVERWRITE=1
+            shift
+            ;;
+          --force-overwrite)
+            OVERWRITE=1
+            shift
+            ;;
+          *)
+            break
+            ;;
+        esac
+done
+
+if [ $OVERWRITE -ne 1 ] && [ -d "$OUTPUT_DIR" ]; then
+  echo "ERR: Folder ${OUTPUT_DIR} already exists. Delete it manually before running this script."
+  exit 1
+fi
+
+mkdir -p "${OUTPUT_DIR}"
+
+# Generate cluster resources.
+kustomize build "${SOURCE_DIR}/cluster" | envsubst > "${CLUSTER_GENERATED_FILE}"
+echo "Generated ${CLUSTER_GENERATED_FILE}"
+
+# Generate controlplane resources.
+kustomize build "${SOURCE_DIR}/controlplane" | envsubst > "${CONTROLPLANE_GENERATED_FILE}"
+echo "Generated ${CONTROLPLANE_GENERATED_FILE}"
+
+# Generate Cluster API provider components file.
+kustomize build "github.com/kubernetes-sigs/cluster-api//config/default/?ref=master" > "${COMPONENTS_CLUSTER_API_GENERATED_FILE}"
+echo "Generated ${COMPONENTS_CLUSTER_API_GENERATED_FILE}"
+
+# Generate Kubeadm Bootstrap Provider components file.
+kustomize build "github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm//config/default/?ref=master" > "${COMPONENTS_KUBEADM_GENERATED_FILE}"
+echo "Generated ${COMPONENTS_KUBEADM_GENERATED_FILE}"
+
+# Generate AWS Credentials.
+AWS_B64ENCODED_CREDENTIALS="$(${CLUSTERAWSADM} alpha bootstrap encode-aws-credentials)"
```

Comment:
This fails if `AWS_REGION` is unset. Should we add a check for this at the top of this script?

Reply:
`clusterawsadm` prints an error if it cannot generate find credentials, that should be enough

Reply:
It took several seconds of running before it failed for me. I'll leave it up to you.

Reply:
Got it, I moved clusterawsadm as the first thing to generate. The main time is spent fetching the kustomize remote repos (CAPI, and kubeadm bootstrapper)